src/repo.rs

```diff
@@ -490,43 +428,175 @@ impl<T: NetworkAdapter + 'static> Repo<T> {
-                                self.process_outgoing_network_messages(&collection_id);
+                    recv(self.control_receiver) -> control_event => {
+                        match control_event {
+                            Ok(ControlEvent::CreateCollection{observer, reply}) => {
+                                let collection = self.new_collection(observer);
+                                reply.send(collection).unwrap();
                             },
-                            Ok(WakeSignal::Sink(collection_id)) => {
-                                self.process_outgoing_network_messages(&collection_id);
+                            Ok(ControlEvent::Shutdown) => {
+                                tracing::info!("shutting down repo");
+                                break;
+                            },
+                            Err(e) => {
+                                tracing::error!("Error receiving control event: {:?}", e);
                             }
                         }
                     },
                 }
             }
         })
     }
+
+    pub fn handle(&self) -> RepoHandle {
+        RepoHandle {
+            control_sender: self.control_sender.clone().expect("Repo has shut down."),
+            collection_sender: self.collection_sender.clone().expect("Repo has shut down."),
+            outboxes: self.outboxes.clone(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct RepoHandle {
+    control_sender: Sender<ControlEvent>,
+    collection_sender: Sender<(CollectionId, CollectionEvent)>,
+    outboxes: Arc<Mutex<HashMap<CollectionId, Outbox>>>,
+}
+
+impl RepoHandle {
+    pub fn create_collection(
+        &self,
+        observer: Option<Box<dyn Fn(Vec<DocumentId>) + Send>>,
+    ) -> DocCollection {
+        let (reply, receiver) = bounded(1);
+        self.control_sender
+            .send(ControlEvent::CreateCollection { observer, reply })
+            .unwrap();
+        receiver.recv().unwrap()
+    }
+
+    pub async fn connect_stream<RecvErr, SendErr, S>(
+        &self,
+        collection: CollectionId,
+        stream: S,
+    ) -> Result<(), error::RunConnection<SendErr, RecvErr>>
+    where
+        RecvErr: std::fmt::Debug,
+        SendErr: std::error::Error,
+        S: Sink<NetworkMessage, Error = SendErr> + Stream<Item = Result<NetworkMessage, RecvErr>>,
+    {
+        let (mut sink, stream) = stream.split();
+        let to_send = self.outgoing(collection).map(|m| Ok::<_, SendErr>(m));
+        let sending = to_send.forward(&mut sink).fuse();
+
+        let recv_msg = self.collection_sender.clone();
+        let recving = stream
+            .map_err(error::RunConnection::<SendErr, RecvErr>::Recv)
+            .try_for_each({
+                move |m| {
+                    futures::future::ready(
+                        recv_msg
+                            .try_send((collection, CollectionEvent::Network(m.into())))
```

Comment:
I don't think this is dropping the message, it's returning an error from `connect_stream` via `TryStreamExt::try_for_each`. At least, that was the intent, if that's not what's happening then yeah this would be a bug.

Returning an error means that it's up to the caller what to do. In the example `run_peer` application this would cause the whole process to end, but you could also implement retry logic there. we could also add some retry logic internally, but to me it generally seems better to do retries at the boundary of the system as that's where you have the information to actually decide whether a retry makes sense.

Reply:
Yes `try_send` returns an error with the original message inside of it.